zebrad/src/components/inbound/downloads.rs

```diff
@@ -217,6 +236,75 @@ where
             };
             metrics::counter!("gossip.downloaded.block.count", 1);
 
+            // # Security & Performance
+            //
+            // Reject blocks that are too far ahead of our tip,
+            // and blocks that are behind the finalized tip.
+            //
+            // Avoids denial of service attacks. Also reduces wasted work on high blocks
+            // that will timeout before being verified, and low blocks that can never be finalized.
+            let tip_height = latest_chain_tip.best_tip_height();
+
+            let max_lookahead_height = if let Some(tip_height) = tip_height {
+                let lookahead = i32::try_from(MAX_INBOUND_CONCURRENCY).expect("fits in i32");
+                (tip_height + lookahead).expect("tip is much lower than Height::MAX")
+            } else {
+                let genesis_lookahead =
+                    u32::try_from(MAX_INBOUND_CONCURRENCY - 1).expect("fits in u32");
+                block::Height(genesis_lookahead)
+            };
+
+            // Get the finalized tip height, assuming we're using the non-finalized state.
+            //
+            // It doesn't matter if we're a few blocks off here, because blocks this low
+            // are part of a fork with much less work. So they would be rejected anyway.
+            //
+            // And if we're still checkpointing, the checkpointer will reject blocks behind
+            // the finalized tip anyway.
+            //
+            // TODO: get the actual finalized tip height
+            let min_accepted_height = tip_height
+                .map(|tip_height| {
+                    block::Height(tip_height.0.saturating_sub(zs::MAX_BLOCK_REORG_HEIGHT))
+                })
+                .unwrap_or(block::Height(0));
+
+            if let Some(block_height) = block.coinbase_height() {
+                if block_height > max_lookahead_height {
+                    tracing::info!(
+                        ?hash,
+                        ?block_height,
+                        ?tip_height,
+                        ?max_lookahead_height,
+                        lookahead_limit = ?MAX_INBOUND_CONCURRENCY,
+                        "gossiped block height too far ahead of the tip: dropped downloaded block"
+                    );
+                    metrics::counter!("gossip.max.height.limit.dropped.block.count", 1);
+
+                    Err("gossiped block height too far ahead")?;
+                } else if block_height < min_accepted_height {
+                    tracing::info!(
```

Comment:
Should these be `debug`? They should happen occasionally, and I'm not sure if there's much a user can do.

That said, it might be hard to distinguish if there's some sort of DoS attack happening if it's hidden in `debug` messages :/

Reply:
I'm happy to make dropping a block that we already have into a debug, because it's impossible for it to cause a hang.

I want to keep dropping a far-future block as an `info`, just in case our limit is wrong, and it causes a hang. That way, it will be obvious in the CI logs.

```suggestion
                    tracing::info!(
                        ?hash,
                        ?block_height,
                        ?tip_height,
                        ?max_lookahead_height,
                        lookahead_limit = ?MAX_INBOUND_CONCURRENCY,
                        "gossiped block height too far ahead of the tip: dropped downloaded block"
                    );
                    metrics::counter!("gossip.max.height.limit.dropped.block.count", 1);

                    Err("gossiped block height too far ahead")?;
                } else if block_height < min_accepted_height {
                    tracing::debug!(
```